src/main/java/org/candlepin/service/impl/PromotedContent.java

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright (c) 2009 - 2021 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+
+package org.candlepin.service.impl;
+
+import org.candlepin.model.Environment;
+import org.candlepin.model.EnvironmentContent;
+import org.candlepin.model.ProductContent;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * A collection of content promoted to environments.
+ */
+public class PromotedContent {
+
+    private static final Logger log = LoggerFactory.getLogger(PromotedContent.class);
+
+    /**
+     * Static factory which collects promoted content from given environments.
+     *
+     * Expects environments to be sorted by priority in order to deduplicate content.
+     *
+     * @param environments environments from which to collect promoted content
+     * @return an instance of {@link ProductContent}
+     */
+    public static PromotedContent from(List<Environment> environments) {
+        if (environments == null || environments.isEmpty()) {
+            return new PromotedContent();
+        }
+
+        log.debug("Consumer has an environment; checking for promoted content.");
+
+        Map<String, EnvironmentContent> promoted = new HashMap<>();
+        for (Environment environment : environments) {
+
+            log.debug("Checking for promoted content in: {}", environment);
+            for (EnvironmentContent envContent : environment.getEnvironmentContent()) {
+                log.debug("  promoted content: {}", envContent.getContent());
+
+                promoted.putIfAbsent(envContent.getContent().getId(), envContent);
+            }
+        }
+
+        return new PromotedContent(promoted);
+    }
+
+    private final Map<String, EnvironmentContent> contents;
+
+    public PromotedContent() {
+        this(new HashMap<>());
+    }
+
+    public PromotedContent(Map<String, EnvironmentContent> contents) {
+        this.contents = contents;
+    }
+
+    /**
+     * Returns an id of environment to which is the given {@link ProductContent} promoted.
+     *
+     * @param pc a product content for which to find environment
+     * @return an environment id
+     */
+    public String environmentOf(ProductContent pc) {
```

Comment:
Should be `environmentIdOf` since we don't actually get the whole environment.

Reply:
Done.